src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/20240802123234_Routine_RebuildIndexes.sql

```diff
@@ -0,0 +1,283 @@
+                   JOIN sys.schemas sch ON sch.schema_id = obj.schema_id
+            WHERE ips.alloc_unit_type_desc = 'IN_ROW_DATA'
+              AND tl.ObjectName = @ObjectName
+              AND sch.Name = @SchemaName
+              AND idx.name = @IndexName) AS Fragmented
+      WHERE __RebuildIndexesAlterIndexes.RunId = @RunId
+        AND __RebuildIndexesAlterIndexes.Id = @StatsRowIndex;
+
+      -- Reorganise or rebuild
+      RAISERROR ('Executing command ''%s''.', 0, 1, @Command) WITH NOWAIT;
+      EXEC (@Command);
+
+      -- Save end info for reporting
+      UPDATE __RebuildIndexesAlterIndexes
+      SET __RebuildIndexesAlterIndexes.EndTime        = GETUTCDATE(),
+          __RebuildIndexesAlterIndexes.EndFragPercent = Fragmented.FragPercent,
+          __RebuildIndexesAlterIndexes.Info           = IIF(@FragPercentBefore >= @FragmentationThresholdRebuild,
+                                                            'Rebuilt', 'Reorganized')
+      FROM (SELECT ips.avg_fragmentation_in_percent AS FragPercent
+            FROM #TableList AS tl
+                   JOIN sys.dm_db_index_physical_stats(DB_ID(), NULL, NULL, NULL, 'LIMITED') ips
+                        ON ips.object_id = OBJECT_ID(tl.ObjectName, 'U')
+                   JOIN sys.indexes idx ON idx.object_id = ips.object_id AND idx.index_id = ips.index_id
+                   JOIN sys.objects obj ON obj.object_id = ips.object_id
+                   JOIN sys.schemas sch ON sch.schema_id = obj.schema_id
+            WHERE ips.alloc_unit_type_desc = 'IN_ROW_DATA'
+              AND tl.ObjectName = @ObjectName
+              AND sch.Name = @SchemaName
+              AND idx.name = @IndexName) AS Fragmented
+      WHERE __RebuildIndexesAlterIndexes.RunId = @RunId
+        AND __RebuildIndexesAlterIndexes.Id = @StatsRowIndex;
+
+      IF @StopTime < GETUTCDATE()
+        BEGIN
+          -- Update Info for all unprocessed indexes, because we're skipping them now we've hit @StopTime
+          UPDATE __RebuildIndexesAlterIndexes
+          SET Info = 'Skipped: stored procedure run time exceeded timeout'
+          WHERE Info IS NULL;
+
+          SET @SkipRemainingTasks = 1;
+        END
+
+      IF @StatsRowIndex >= @StatsCount
+        RAISERROR ('Completed optimising indexes for tables.', 0, 1, NULL) WITH NOWAIT;
+
+      SET @StatsRowIndex += 1;
+    END
+
+  /*-----------------------------------------------------------------------*/
+
+  SET @TablesRowIndex = 1;
+  SELECT @TablesCount = COUNT(Id) FROM #TableList;
+  WHILE @TablesRowIndex <= @TablesCount AND @SkipRemainingTasks = 0
+    BEGIN
+      DECLARE @CurrentTable NVARCHAR(MAX) = '';
+      SELECT @CurrentTable = ObjectName
+      FROM #TableList
+      WHERE Id = @TablesRowIndex;
+
+      SET @Command = 'UPDATE STATISTICS ' + @CurrentTable;
+
+      -- Save start time
+      UPDATE __RebuildIndexesUpdateStatistics
+      SET StartTime = GETUTCDATE()
+      WHERE RunId = @RunId
+        AND TableName = @CurrentTable;
```

Comment:
Because there's no other data like `FragPercent` kept on this table, maybe you don't need to insert rows into this table early and update them with `StartTime` here, but put the insert here instead?

Saying that though, for the previous stage you have a method of updating each row when the timeout is hit which you don't have here. So maybe `__RebuildIndexesUpdateStatistics` should get a `HitTimeout` column which gets updated here on timeout, in which case keep the original method of inserting the rows early! 🙂

Reply:
I'm sure you're aware, but to make it clear, as there is no frag percent threshold with update statistics, there is no ambiguity as to why those rows are skipped if they have no start time - it's because of the timeout.

The action I'm most tempted to take would be to add a HitTimeout column, just so it makes it absolutely clear what has happened to anyone looking at the `__Log_RebuildIndexesUpdateStatistics` table with no other context about how it works.

I'm inclined to keep creating the rows early and not add a HitTimeout column - it will mean we execute one extra query, but I don't think that's a big deal. I think it makes sense for readability to follow the same pattern as the rebuilding rather than have two different approaches. And we avoid having to add an extra INSERT into `IF @TablesRowIndex + 1 <= @TablesCount AND @StopTime < GETUTCDATE()`.

All that said, I'm inclined to think this is all pretty minor and think any benefit of changing this isn't worth the time to do so.

Reply:
I think it's fine to leave this.